modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/CurrencyMessageValidator.scala

```diff
@@ -0,0 +1,113 @@
+    implicit hasher: Hasher[F]
+  ): F[CurrencyMessageOrError]
+}
+
+object CurrencyMessageValidator {
+
+  def make[F[_]: Async: SecurityProvider](
+    validator: SignedValidator[F],
+    allowanceList: Option[Map[Address, NonEmptySet[PeerId]]],
+    seedlist: Option[Set[SeedlistEntry]]
+  ): CurrencyMessageValidator[F] =
+    new CurrencyMessageValidator[F] {
+      def validate(
+        message: Signed[CurrencyMessage],
+        lastMessages: SortedMap[MessageType, Signed[CurrencyMessage]],
+        metagraphId: Address
+      )(implicit hasher: Hasher[F]): F[CurrencyMessageOrError] = {
+
+        val allowancePeers = allowanceList
+          .map(
+            _.get(metagraphId)
+              .map(_.toSortedSet)
+              .getOrElse(SortedSet.empty[PeerId])
+          )
+
+        val seedlistPeers = seedlist.map(_.map(_.peerId))
+
+        val isMetagraphIdValid = metagraphId === message.metagraphId
+
+        def validateIfSignedBySeedlistPeersExcludingOwner(message: Signed[CurrencyMessage]) =
+          message.proofs.toList
+            .traverse(sp => sp.id.toAddress.map((sp.id, _)))
+            .map(_.collectFirst { case (id, addr) if addr === message.address => id })
+            .map(_.flatMap { ownerId =>
+              NonEmptySet
+                .fromSet(message.proofs.filter(_.id =!= ownerId))
+                .map(sigs => Signed(message.value, sigs))
+            })
+            .map {
+              case Some(noOwnerSigMsg) =>
+                validator
+                  .validateSignaturesWithSeedlist(seedlistPeers, noOwnerSigMsg)
+                  .map(_ => message)
+              case None => Validated.validNec(message)
+            }
+
+        def validateSignatures(message: Signed[CurrencyMessage]) =
+          for {
+            isSignedByOwner <- validator.isSignedBy(message, message.address)
+            hasNoDuplicates = validator.validateUniqueSigners(message)
+            isSignedCorrectly <- validator.validateSignatures(message)
+            isSignedByMajority = validator.validateSignedBySeedlistMajority(allowancePeers, message)
+            isSignedBySeedlistPeers <- validateIfSignedBySeedlistPeersExcludingOwner(message)
+          } yield
+            isSignedByOwner
+              .productR(hasNoDuplicates)
+              .productR(isSignedCorrectly)
+              .productR(isSignedByMajority)
+              .productR(isSignedBySeedlistPeers)
+
+        lastMessages.get(message.messageType) match {
+          case Some(lastMessage) if message.parentOrdinal =!= lastMessage.ordinal =>
+            Async[F].pure(NotANextMessage.invalidNec)
+          case None if message.parentOrdinal =!= MessageOrdinal.MinValue =>
+            Async[F].pure(FirstMessageWithWrongOrdinal.invalidNec)
+          case _ if !isMetagraphIdValid =>
+            Async[F].pure(WrongMetagraphId.invalidNec)
+          case _ =>
+            validateSignatures(message).map(mapError)
```

Comment:
```suggestion
            validateSignatures(message)
```
... if we put the `leftMap` up above